provider/lxd/environ.go

```diff
@@ -365,3 +373,45 @@ func (env *environ) ReplaceOrAddInstanceProfile(instId, oldProfile, newProfile s
 	}
 	return env.LXDProfileNames(instId)
 }
+
+// AssignProfiles implements environs.LXDProfiler.
+func (env *environ) AssignProfiles(instId string, profilesNames []string, profilePosts []lxdprofile.ProfilePost) (current []string, err error) {
+	defer func() ([]string, error) {
+		if err != nil {
+			var err2 error
+			current, err2 = env.LXDProfileNames(instId)
+			if err2 != nil {
+				logger.Errorf("secondary error, retrieving profile names: %s", err2)
+			}
+		}
+		return current, err
+	}()
+
+	deleteProfiles := []string{}
+
+	// Write any new profilePosts and gather a slice of profile
+	// names to be deleted, after removal.
+	for _, p := range profilePosts {
+		if p.Profile != nil {
+			pr := charm.LXDProfile(*p.Profile)
+			if err := env.MaybeWriteLXDProfile(p.Name, &pr); err != nil {
+				return nil, err
+			}
+		} else {
+			deleteProfiles = append(deleteProfiles, p.Name)
+		}
+	}
+
+	server := env.server()
+	if err := server.UpdateContainerProfiles(instId, profilesNames); err != nil {
+		return []string{}, errors.Trace(err)
+	}
+
+	for _, name := range deleteProfiles {
+		if err := server.DeleteProfile(name); err != nil {
+			// most likely the failure is because the profile is already in use
```

Comment:
Orphan profiles although look bad from the point of view of `lxd profile list` isn't, unless it grows to 1000s I guess. We should make some sort of documentation around this.

Reply:
The last profile on a machine destroying may not be deleted.  There might be a way with our new worker, to attempt to delete them. Will have to ponder.